Break out ECS task metadata clients into ecsutil

rmfitzpatrick opened the pull request:
**Description:**
These changes unify the two ECS task metadata client implementations from the ECS Resource Detection Processor and the AWS ECS Container Metrics Receiver.  These are intended to be completely transparent but there are minor implementation detail and interface updates around metadata endpoint detection (and its error reporting) and the client/client provider types from the container metrics receiver.  Container stats remain in the metric receiver since they are not used by the resource detector.  Adoption of core's HTTPClientSettings is also included per a TODO comment in the receiver.

These changes are also intended to assist the implementation of an ECS sidecar watch observer as mentioned in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5316.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3188

**Testing:**
Existing tests updated to meet helper relocation and updated for new helpers.  I also manually tested both the resource detector and container metrics receiver in an ECS cluster.

**Documentation:**
No documentation is currently provided with this internal package, as appears to be common.

cc: @alolita

MovieStoreGuy commented on internal/aws/ecsutil/client_test.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-package awsecscontainermetrics
+package ecsutil
```
the `aws` naming should be kept with all modules committed

```suggestion
package awsescutil
```

rmfitzpatrick replied:
None of the other aws internal packages follow this suggested convention (are all in an `aws` directory).

MovieStoreGuy commented on internal/aws/ecsutil/doc.go:
```diff
@@ -0,0 +1,17 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ecsutil provides a common task metadata definition and client for ECS environments
```
```suggestion
// Package awsecsutil provides a common task metadata definition and client for ECS environments
package awsecsutil
```

rmfitzpatrick replied:
same as https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5994#discussion_r740208724

MovieStoreGuy commented on internal/aws/ecsutil/metadata_endpoints.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// GetTMEFromEnv will return the first available task metadata endpoint for the v4 or v3 env var in that order.
+func GetTMEFromEnv() (*url.URL, error) {
```
```suggestion
func GetTMEFromEnv() (endpoint *url.URL, err error) {
```

rmfitzpatrick replied:
updated all applicable helpers

MovieStoreGuy commented on internal/aws/ecsutil/metadata_provider.go:
```diff
@@ -0,0 +1,99 @@
+	FetchContainerMetadata() (*ContainerMetadata, error)
+}
+
+type ECSMetadataProviderImpl struct {
```
This does not need to exported since there is a function to return a Provider and the zero value would cause a panic.

rmfitzpatrick replied:
Thanks, updated.

MovieStoreGuy commented on internal/aws/ecsutil/metadata_provider.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+type ECSMetadataProviderImpl struct {
+	Logger *zap.Logger
```
The internal variables don't need to be exported as well.

MovieStoreGuy reviewed (changes requested):
I have a question about how this is different than from https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/detectors/aws/ecs ? 

It looks like we are doing the same work, but duplicating the means of getting there. Does that sound right?

MovieStoreGuy commented:
Sorry, I had gotten half way through a noticed it looked similar to what the resource detector was doing.

MovieStoreGuy commented on receiver/awsecscontainermetricsreceiver/receiver_test.go:
```diff
@@ -17,28 +17,38 @@ package awsecscontainermetricsreceiver
-	taskStats, err := ioutil.ReadFile("testdata/task_stats.json")
-	if err != nil {
-		return nil, nil, err
+func (f fakeRestClient) GetResponse(path string) ([]byte, error) {
```
I suggest using `embed` instead of reading the files each time here since they would be in memory already and would help speed up tests

rmfitzpatrick replied:
I think this is the standard method of handling `testdata` files in the project and don't believe embed is used anywhere else.  If it were part of the application I can see that being helpful though.

MovieStoreGuy replied:
Yeah, that is reasonable :)

rmfitzpatrick replied:
I actually found this necessary after consolidating a test helper and am now a convert.  Thanks for the suggestion.

rmfitzpatrick commented:
> I have a question about how this is different than from https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/detectors/aws/ecs ?
> 
> It looks like we are doing the same work, but duplicating the means of getting there. Does that sound right?

I wasn't familiar with the go-contrib detector (doesn't look like the collector uses the go library as a dependency) but from a basic implementation check they aren't overlapping functionally very much.  This util reads from the task metadata endpoint and provides an interface to its task and container contents whereas the go detector gets the `os.Hostname` and parses the container id from the cgroup entries if one of the task metadata env vars is set.

MovieStoreGuy commented on receiver/awsecscontainermetricsreceiver/receiver_test.go:
```diff
@@ -17,28 +17,38 @@ package awsecscontainermetricsreceiver
-	if err != nil {
-		return nil, nil, err
+func (f fakeRestClient) GetResponse(path string) ([]byte, error) {
+	var content []byte
```
Can I suggest you change it to the following to remove the need for error handling?

```suggestion
func (f fakeRestClient) GetResponse(uri string) ([]byte, error) {
    var testfile string
    switch uri {
    case escutil.TaskMetadataPath:
       testfile = "testdata/task_metadata.json"
    case awscontainermetrics.TaskStatsPath:
       testfile = "testdata/task_stats.json"
    }
    return io.ReadFile(testfile)
}
```

rmfitzpatrick replied:
thanks, I've updated with the new helper

anuraaga commented on internal/aws/ecsutil/testdata/task_metadata.json:
```diff
@@ -0,0 +1,109 @@
+{
```
Is there a copy of this file left over that we can delete and reference this one instead?

rmfitzpatrick replied:
I removed the initial from the ecs container metrics receiver and am using the broken out rest client test helper.

anuraaga commented on internal/aws/ecsutil/metadata_provider.go:
```diff
@@ -0,0 +1,99 @@
+
+// FetchContainerMetadata retrieves the metadata for the Amazon ECS Container the collector is running on
+func (md *ecsMetadataProviderImpl) FetchContainerMetadata() (*ContainerMetadata, error) {
+	resp, err := md.client.GetResponse("")
```
Can you extract a constant ContainerMetadataPath instead of magic empty string?

rmfitzpatrick replied:
thanks, updated

anuraaga commented on receiver/awsecscontainermetricsreceiver/receiver_test.go:
```diff
@@ -17,28 +17,38 @@ package awsecscontainermetricsreceiver
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics"
 )
 
 type fakeRestClient struct {
```
Would it make sense to put this in a `ecsutiltest` package that's shared?

anuraaga replied:
The reason this came to mind in particular is that it seems a bit strange for the path variables to be exposed downstream like here

rmfitzpatrick replied:
I took a shot at this with a `testing.T` in the signature to account for exported test util.

anuraaga reviewed (commented):
Just some small points, thanks

MovieStoreGuy commented:
If you rebase against main, it should fix the issues with the prom compliance test as well.

anuraaga commented on internal/aws/ecsutil/testutil.go:
```diff
@@ -0,0 +1,39 @@
+	"testing"
+)
+
+//go:embed testdata/container_metadata.json
```
I believe this will cause this testdata and the testing module to get embedded into the production binary. Is that right? It's why a separate package would be safer I think, though I don't think it needs to be a separate go module.

MovieStoreGuy replied:
Seperate package makes sense to me, something of the name `ecsutiltest` would be clear that it is not intended for production use.

rmfitzpatrick replied:
I looked into this is a bit w/ hexdump and as far as I can tell the testdata ss only embedded in the production binary if the []byte vars are directly used in non-test code.  I still migrated the test helper to its own package though.

rmfitzpatrick commented:
@bogdandrutu any chance you could take a look/help this land when able?

Commit:
Break out ECS task metadata clients into ecsutil

These changes unify the two ECS task metadata client implementations from the ECS Resource Detection Processor
and the AWS ECS Container Metrics Receiver. These are intended to be completely transparent but there are minor
implementation detail and interface updates around metadata endpoint detection (and its error reporting) and the
client/client provider types from the container metrics receiver. Container stats remain in the metric receiver
since they are not used by the resource detector. Adoption of core's HTTPClientSettings is also included per a
TODO comment in the receiver.

These changes are also intended to assist the implementation of an ECS sidecar watch observer as mentioned in #5316.

tigrannajaryan commented:
@rmfitzpatrick just to confirm: this has no user-visible behavior changes, right?

rmfitzpatrick commented:
> @rmfitzpatrick just to confirm: this has no user-visible behavior changes, right?

Right, it's completely transparent to the end user, but there are some new exported items for future use in other components.
